src/chainstate/stacks/index/marf.rs

```diff
@@ -0,0 +1,2187 @@
+                                        storage.open_block(block_hash, true)?;
+                                        return Ok(cursor);
+                                    },
+                                    _ => {
+                                        return Err(e);
+                                    }
+                                }
+                            }
+                        };
+
+                        // finish taking the step
+                        cursor.walk_backptr_finish(&next_node_ptr, &next_node_block_hash);
+                        
+                        node = next_node;
+                        node_ptr = next_node_ptr;
+                        
+                        storage.open_block(block_hash, true)?;
+                    }
+                }
+            }
+        }
+
+        trace!("Trie has a cycle");
+        return Err(Error::CorruptionError("Trie has a cycle".to_string()));
+    }
+
+
+    /// Walk down this MARF at the given block hash, resolving backptrs to previous tries.
+    /// Return the cursor and the last node visited.
+    /// s will point to the block in which the leaf was found, or the last block visited.
+    fn walk(storage: &mut TrieFileStorage, block_hash: &BlockHeaderHash, path: &TriePath) -> Result<(TrieCursor, TrieNodeType), Error> {
+        storage.open_block(block_hash, false)?;
+
+        let root_ptr = storage.root_ptr();
+        let mut cursor = TrieCursor::new(path, root_ptr);
+
+        // walk to insertion point 
+        let (mut node, _) = Trie::read_root(storage)?;
+        let mut node_ptr = TriePtr::new(0,0,0);
+
+        for _ in 0..(cursor.path.len()+1) {
+            let next_opt = Trie::walk_from(storage, &node, &mut cursor)?;
+            match next_opt {
+                Some((next_node_ptr, next_node, _)) => {
+                    // keep walking
+                    node = next_node;
+                    node_ptr = next_node_ptr;
+                    continue;
+                },
+                None => {
+                    if cursor.div() {
+                        // we're done -- path diverged.  No backptr-walking can help us.
+                        trace!("Path diverged -- we're done.");
+                        return Err(Error::NotFoundError);
+                    }
+                    else {
+                        // we're not done with this path.  Either no node exists, or it exists off
+                        // of a prior version of the last-visited node.
+                        let chr = cursor.chr().unwrap();     // guaranteed to succeed since we walked some path.
+                        let found_leaf =
+                            if node.is_leaf() {
+                                if !cursor.eop() {
+                                    // at an existing leaf with a different path.
```

Comment:
is this branch reachable? wouldn't this indicate that the path diverged, since you landed at a leaf with a different path?

Reply:
Ah, you're right.  This shouldn't be reachable.  I'll replace it with an `assert!(cursor.eop())`.

Reply:
Should be fixed now.